src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -159,28 +153,38 @@ internal bool TryAddWithoutValidation(HeaderDescriptor descriptor, IEnumerable<s
                 throw new ArgumentNullException(nameof(values));
             }
 
-            using (IEnumerator<string?> enumerator = values.GetEnumerator())
+            using IEnumerator<string?> enumerator = values.GetEnumerator();
+            if (enumerator.MoveNext())
             {
+                string firstValue = enumerator.Current ?? string.Empty;
                 if (enumerator.MoveNext())
                 {
-                    TryAddWithoutValidation(descriptor, enumerator.Current);
-                    if (enumerator.MoveNext())
+                    var valuesList = new List<string>
```

Comment:
Consider initializing the list capacity with `Enumerable.TryGetNonEnumeratedCount`

Reply:
Actually this library doesn't depend on LINQ, so might not be the best reason to add a dependency.

Reply:
We could also save the allocation for the enumerator here by trying to cast to concrete types first, but I didn't want to add that complexity in this PR.